File: tests/c/test_population.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: LicenseRef-NvidiaProprietary
//
// NVIDIA CORPORATION, its affiliates and licensors retain all intellectual
// property and proprietary rights in and to this material, related
// documentation and any modifications thereto. Any use, reproduction,
// disclosure or distribution of this material and related documentation
// without an express license agreement from NVIDIA CORPORATION or
// its affiliates is strictly prohibited.
//
// Public ovstage population test (USD -> ovstage): populate from inline USDA and
// query back; add/remove a USD reference; reset; and confirm a missing file
// fails. Tested source for the loading-usd skill's C snippets; the Python sibling
// is tests/python/test_population.py. CPU-only.

#include <ovstage/ovstage.h>
#include <ovx/path_dictionary/path_dictionary.h>

#include <gtest/gtest.h>

#include <cmath>
#include <cstring>
#include <filesystem>
#include <fstream>
#include <string>

namespace
{

ovx_string_t str(const char* s)
{
    return ovx_string_t{ s, std::strlen(s) };
}

// Drive an ordinary (non-population) enqueue to completion.
bool waitOp(ovstage_instance_t* stage, ovstage_enqueue_result_t enq)
{
    if (enq.status != OVSTAGE_OK)
        return false;
    ovstage_op_wait_result_t wait{};
    const ovstage_api_status_t status = ovstage_wait_op(stage, enq.op_index, OVSTAGE_TIMEOUT_INFINITE, &wait);
    ovstage_release_op(stage, enq.op_index);
    return status == OVSTAGE_OK && wait.error_op_id_count == 0;
}

// Drive a population enqueue to completion; true iff it ran error-free.
bool waitPopOk(ovstage_instance_t* stage, ovstage_population_enqueue_result_t pop, const char* what)
{
    if (pop.status != OVSTAGE_OK)
    {
        ADD_FAILURE() << what << " enqueue rejected (code " << pop.status << ")";
        return false;
    }
    ovstage_population_op_wait_result_t wait{};
    const ovstage_api_status_t st = ovstage_population_wait_op(stage, pop.op_index, OVSTAGE_TIMEOUT_INFINITE, &wait);
    if (st != OVSTAGE_OK || wait.error_op_id_count != 0)
    {
        ADD_FAILURE() << what << " op failed (code " << st << ")";
        return false;
    }
    return true;
}

// A population op expected to FAIL (rejected at enqueue, or failed at wait).
bool waitPopFails(ovstage_instance_t* stage, ovstage_population_enqueue_result_t pop)
{
    if (pop.status != OVSTAGE_OK)
        return true;
    ovstage_population_op_wait_result_t wait{};
    const ovstage_api_status_t st = ovstage_population_wait_op(stage, pop.op_index, OVSTAGE_TIMEOUT_INFINITE, &wait);
    return st != OVSTAGE_OK || wait.error_op_id_count != 0;
}

// A minimal USD scene: one Cube under a World Xform.
const char kCubeUsda[] =
    "#usda 1.0\n"
    "(\n"
    "    defaultPrim = \"World\"\n"
    ")\n"
    "def Xform \"World\"\n"
    "{\n"
    "    def Cube \"Cube\"\n"
    "    {\n"
    "        double size = 1.0\n"
    "    }\n"
    "}\n";

// A self-contained, referenceable layer (defaultPrim set so the reference composes).
const char kRefUsda[] =
    "#usda 1.0\n"
    "(\n"
    "    defaultPrim = \"Ref\"\n"
    ")\n"
    "def Xform \"Ref\"\n"
    "{\n"
    "    def Cube \"Cube\"\n"
    "    {\n"
    "        double size = 1.0\n"
    "    }\n"
    "}\n";

// A scene where a schema selection is distinguishable from selecting everything.
const char kSchemaUsda[] =
    "#usda 1.0\n"
    "(\n"
    "    defaultPrim = \"World\"\n"
    ")\n"
    "def Xform \"World\"\n"
    "{\n"
    "    def Cube \"Body\" (prepend apiSchemas = [\"PhysicsRigidBodyAPI\"])\n"
    "    {\n"
    "        double size = 4.0\n"
    "    }\n"
    "    def Cube \"Prop\"\n"
    "    {\n"
    "        double size = 2.0\n"
    "    }\n"
    "    def Sphere \"Ball\"\n"
    "    {\n"
    "        double radius = 1.0\n"
    "    }\n"
    "}\n"
    "def Cube \"Outside\"\n"
    "{\n"
    "    double size = 9.0\n"
    "}\n";

class PopulationTest : public ::testing::Test
{
protected:
    void SetUp() override
    {
        ovstage_instance_desc_t desc{};
        desc.name = "test.ovstage.population";
        ASSERT_EQ(ovstage_create_instance(&desc, &stage_), OVSTAGE_OK);
        ASSERT_NE(stage_, nullptr);
    }

    void TearDown() override
    {
        if (stage_)
            ovstage_destroy_instance(stage_);
    }

    // Count prims whose usd-path starts with `prefix` (latest committed state).
    size_t prefixCount(const char* prefix)
    {
        ovx_string_t value = str(prefix);
        ovstage_predicate_t pred{};
        pred.attribute.string = str("usd-path");
        pred.op = OVSTAGE_FILTER_OP_PREFIX;
        pred.values = &value;
        pred.value_count = 1;
        ovstage_filter_t filter{};
        filter.predicates = &pred;
        filter.count = 1;

        ovstage_query_handle_t query = OVSTAGE_INVALID_QUERY_HANDLE;
        if (!waitOp(stage_, ovstage_query(stage_, &filter, nullptr, 0, &query)))
        {
            if (query != OVSTAGE_INVALID_QUERY_HANDLE)
                waitOp(stage_, ovstage_release_query(stage_, query));
            return 0;
        }
        ovstage_query_result_t result{};
        size_t count = 0;
        if (ovstage_fetch_query_result(stage_, query, OVSTAGE_TIMEOUT_INFINITE, &result) == OVSTAGE_OK)
        {
            count = result.total_prim_count;
            ovstage_release_query_result(stage_, &result);
        }
        waitOp(stage_, ovstage_release_query(stage_, query));
        return count;
    }

    ovstage_instance_t* stage_ = nullptr;
};

TEST_F(PopulationTest, PopulateFromUsdaAndQuery)
{
    size_t matched = 0;
    ovstage_query_handle_t query = OVSTAGE_INVALID_QUERY_HANDLE;

    // [snippet:populate-and-query-c]
    // Populate the ovstage from an inline USDA string (the RENDERING domain
    // mirrors meshes/lights/materials/cameras). Population is asynchronous — wait
    // for the op. Then confirm the prim landed by querying it back by its
    // usd-path; queries resolve against the latest committed state.
    ovstage_population_enqueue_result_t pop = ovstage_population_open_usd_from_string(
        stage_, str(kCubeUsda), /*ordinal=*/1, /*time=*/0.0, OVSTAGE_POPULATION_DOMAIN_RENDERING);
    ASSERT_EQ(pop.status, OVSTAGE_OK);
    ASSERT_EQ(ovstage_population_wait_op(stage_, pop.op_index, OVSTAGE_TIMEOUT_INFINITE, nullptr), OVSTAGE_OK);

    ovx_string_t pathValue = str("/World/Cube");
    ovstage_predicate_t predicate{};
    predicate.attribute.string = str("usd-path");
    predicate.op = OVSTAGE_FILTER_OP_IN;
    predicate.values = &pathValue;
    predicate.value_count = 1;

    ovstage_filter_t filter{};
    filter.predicates = &predicate;
    filter.count = 1;

    const bool queryOk = waitOp(stage_, ovstage_query(stage_, &filter, nullptr, 0, &query));
    if (!queryOk)
    {
        if (query != OVSTAGE_INVALID_QUERY_HANDLE)
            waitOp(stage_, ovstage_release_query(stage_, query));
        ASSERT_TRUE(queryOk);
    }

    ovstage_query_result_t result{};
    const ovstage_api_status_t fetched = ovstage_fetch_query_result(stage_, query, OVSTAGE_TIMEOUT_INFINITE, &result);
    if (fetched == OVSTAGE_OK)
    {
        matched = result.total_prim_count;
        ovstage_release_query_result(stage_, &result);
    }
    // [/snippet:populate-and-query-c]

    ASSERT_TRUE(waitOp(stage_, ovstage_release_query(stage_, query)));
    ASSERT_EQ(fetched, OVSTAGE_OK);

    EXPECT_EQ(matched, 1u);
}

TEST_F(PopulationTest, AddRemoveUsdReference)
{
    ASSERT_TRUE(waitPopOk(stage_,
        ovstage_population_open_usd_from_string(stage_, str(kCubeUsda), 1, NAN, OVSTAGE_POPULATION_DOMAIN_RENDERING),
        "open"));

    // [snippet:usd-reference-c]
    // add_usd_reference edits the USD source only; apply_usd_changes propagates it
    // into the stage (at the ordinal you pass). The add reserves a handle
    // synchronously for a later remove; removing again propagates the tombstone.
    ovstage_population_usd_reference_handle_t handle = OVSTAGE_POPULATION_INVALID_USD_REFERENCE_HANDLE;
    ASSERT_TRUE(waitPopOk(stage_,
        ovstage_population_add_usd_reference_from_string(stage_, str(kRefUsda), str("/World/Props"), &handle),
        "add_reference"));
    ASSERT_TRUE(waitPopOk(stage_, ovstage_population_apply_usd_changes(stage_, 2), "apply_usd_changes"));
    EXPECT_GT(prefixCount("/World/Props"), 0u);  // the referenced subtree materialized

    ASSERT_TRUE(waitPopOk(stage_, ovstage_population_remove_usd_reference(stage_, handle), "remove_reference"));
    ASSERT_TRUE(waitPopOk(stage_, ovstage_population_apply_usd_changes(stage_, 3), "apply_usd_changes"));
    EXPECT_EQ(prefixCount("/World/Props"), 0u);  // the referenced subtree was removed
    // [/snippet:usd-reference-c]

    // The handle is spent: removing it again fails.
    EXPECT_TRUE(waitPopFails(stage_, ovstage_population_remove_usd_reference(stage_, handle)));
}

TEST_F(PopulationTest, ResetUsdAndRepopulate)
{
    ASSERT_TRUE(waitPopOk(stage_,
        ovstage_population_open_usd_from_string(stage_, str(kCubeUsda), 1, NAN, OVSTAGE_POPULATION_DOMAIN_RENDERING),
        "open"));
    EXPECT_GT(prefixCount("/World"), 0u);

    // [snippet:reset-usd-c]
    // reset_usd clears the USD source; apply_usd_changes propagates the cleared
    // state. The stage stays usable afterwards — repopulating from USD still works.
    ASSERT_TRUE(waitPopOk(stage_, ovstage_population_reset_usd(stage_), "reset_usd"));
    ASSERT_TRUE(waitPopOk(stage_, ovstage_population_apply_usd_changes(stage_, 2), "apply_usd_changes"));
    EXPECT_EQ(prefixCount("/World"), 0u);  // cleared before repopulating
    ASSERT_TRUE(waitPopOk(stage_,
        ovstage_population_open_usd_from_string(stage_, str(kCubeUsda), 3, NAN, OVSTAGE_POPULATION_DOMAIN_RENDERING),
        "reopen"));
    // [/snippet:reset-usd-c]

    EXPECT_GT(prefixCount("/World"), 0u);
}

TEST_F(PopulationTest, OpenMissingFileFails)
{
    // [snippet:open-missing-file-c]
    // A missing/unreadable file fails the populate op: the enqueue is accepted but
    // the failure surfaces from ovstage_population_wait_op.
    EXPECT_TRUE(waitPopFails(stage_,
        ovstage_population_open_usd_from_file(
            stage_, str("/nonexistent/ovstage-does-not-exist.usda"), 1, NAN, OVSTAGE_POPULATION_DOMAIN_RENDERING)));
    // [/snippet:open-missing-file-c]
}

} // namespace

// Registration is process-global, irreversible, and only counts before USD first reads
// its schema definitions. gtest_discover_tests gives every case its own process, so this
// one registers into a fresh USD and leaves nothing behind for the others.
TEST_F(PopulationTest, RegisterUsdSchemas)
{
    // A single-family schema tree USD has never seen -- a core Usd* schema would resolve
    // whether or not the call below was made, and so would prove nothing. The family
    // declares one fallback, which is what the assertion at the end reads back.
    const std::filesystem::path root =
        std::filesystem::temp_directory_path() / "ovstage_public_schema_fixture";
    const std::filesystem::path resources = root / "resources";
    std::filesystem::remove_all(root);
    std::filesystem::create_directories(resources);

    std::ofstream descriptor((resources / "plugInfo.json").string(), std::ios::binary | std::ios::trunc);
    descriptor << R"({
    "Plugins": [
        {
            "Info": { "Types": {
                "OvstagePublicTestAPI": {
                    "alias": { "UsdSchemaBase": "OvstagePublicTestAPI" },
                    "bases": [ "UsdAPISchemaBase" ],
                    "schemaIdentifier": "OvstagePublicTestAPI",
                    "schemaKind": "singleApplyAPI"
                }
            } },
            "Name": "ovstagePublicTestSchema",
            "ResourcePath": "resources",
            "Root": "..",
            "Type": "resource"
        }
    ]
}
)";
    descriptor.close();
    ASSERT_TRUE(descriptor);

    std::ofstream generated((resources / "generatedSchema.usda").string(), std::ios::binary | std::ios::trunc);
    generated << "#usda 1.0\n(\n)\n\nclass \"OvstagePublicTestAPI\" (\n)\n{\n"
              << "    uniform float ovstageTest:strength = 1.5\n}\n";
    generated.close();
    ASSERT_TRUE(generated);

    // A prim applying that family and authoring nothing of it.
    static const char kUsda[] =
        "#usda 1.0\n"
        "(\n"
        "    defaultPrim = \"World\"\n"
        ")\n"
        "def Xform \"World\"\n"
        "{\n"
        "    def Cube \"Body\" (prepend apiSchemas = [\"OvstagePublicTestAPI\"])\n"
        "    {\n"
        "        double size = 4.0\n"
        "    }\n"
        "}\n";

    const std::string schemaDir = resources.string();

    // [snippet:register-usd-schemas-c]
    // ovstage registers only what its USD build provides (the core Usd* schemas).
    // Register any other family your stages use -- each path a plugInfo.json or a
    // directory holding one -- before the first ovstage call that reads schema
    // definitions, which population and export both do: USD reads them once, and a
    // family registered afterwards contributes nothing.
    ovx_string_t schemaPaths[] = { str(schemaDir.c_str()) };
    ASSERT_EQ(ovstage_population_register_usd_schemas(schemaPaths, 1), OVSTAGE_OK);

    // With the family registered, a description can select on it and publish what it
    // declares -- including properties the stage never authored, which resolve to their
    // schema fallbacks. Without the registration they would not exist to publish.
    ovx_string_t family[] = { str("OvstagePublicTestAPI") };
    ovstage_population_selector_t selector = {};
    selector.prim_predicate = ovstage_population_prim_predicate_has_applied_schema(family, 1);
    selector.property_predicate = ovstage_population_property_predicate_declared_by_schema(family, 1);
    ovstage_population_desc_t desc = {};
    desc.selectors = &selector;
    desc.selector_count = 1;

    ASSERT_TRUE(waitPopOk(stage_,
        ovstage_population_open_usd_from_string_with_desc(
            stage_, str(kUsda), /*ordinal=*/1, /*time=*/NAN, &desc, 1),
        "populate with a registered schema family"));
    // [/snippet:register-usd-schemas-c]

    // Read the unauthored attribute back. The stage carries no such value, so it
    // resolves only because the family's definitions were registered.
    ovstage_write_floor_desc_t writeFloor{};
    writeFloor.ordinal = 1;
    writeFloor.scope = OVSTAGE_SCOPE_ALL;
    ASSERT_TRUE(waitOp(stage_, ovstage_advance_write_floor(stage_, &writeFloor)));

    ovx_string_t primPath = str("/World/Body");
    ovstage_predicate_t pred{};
    pred.attribute.string = str("usd-path");
    pred.op = OVSTAGE_FILTER_OP_IN;
    pred.values = &primPath;
    pred.value_count = 1;
    ovstage_filter_t filter{};
    filter.predicates = &pred;
    filter.count = 1;
    ovstage_query_handle_t query = OVSTAGE_INVALID_QUERY_HANDLE;
    ASSERT_TRUE(waitOp(stage_, ovstage_query(stage_, &filter, nullptr, 0, &query)));

    path_dictionary_instance_t* dict = ovstage_get_path_dictionary(stage_);
    ASSERT_NE(dict, nullptr);
    ovx_string_t attrName = str("ovstageTest:strength");
    ovx_token_t attr = OVX_INVALID_TOKEN;
    ASSERT_EQ(path_dictionary_create_tokens_from_strings(dict, &attrName, 1, &attr).status,
              OVX_API_SUCCESS);

    ovstage_ordinal_range_t range{};
    range.end_ordinal = 1;
    range.has_start_ordinal = false;
    ovstage_read_handle_t read = OVSTAGE_INVALID_READ_HANDLE;
    ASSERT_TRUE(waitOp(stage_, ovstage_read_attributes(stage_, query, &attr, 1, range, &read)));

    ovstage_read_group_t group{};
    ASSERT_EQ(ovstage_fetch_read_next(stage_, read, OVSTAGE_TIMEOUT_INFINITE, &group), OVSTAGE_OK);
    ASSERT_EQ(group.data.tensor_count, 1u);
    ASSERT_NE(group.data.tensors[0].data, nullptr);
    float strength = NAN;
    std::memcpy(&strength, group.data.tensors[0].data, sizeof(strength));
    EXPECT_FLOAT_EQ(strength, 1.5f);

    ovstage_release_group(stage_, &group);
    waitOp(stage_, ovstage_release_read(stage_, read));
    waitOp(stage_, ovstage_release_query(stage_, query));
    std::filesystem::remove_all(root);
}

// Populate exactly the prims carrying a schema, with the properties that schema declares.
TEST_F(PopulationTest, PopulateBySchema)
{
    // [snippet:populate-by-schema-c]
    // A description says what to populate. The prim predicate picks the prims -- here
    // anything conforming to PhysicsRigidBodyAPI, whether as its type or as an applied
    // API schema -- and the property predicate picks which of their properties are
    // published. The same schema list serves both halves.
    //
    // A predicate references its values rather than copying them, so `schemas` must
    // outlive the call; the description itself need not outlive it.
    ovx_string_t schemas[] = { str("PhysicsRigidBodyAPI") };

    ovstage_population_selector_t selector = {};
    selector.prim_predicate = ovstage_population_prim_predicate_has_schema(schemas, 1);
    selector.property_predicate = ovstage_population_property_predicate_declared_by_schema(schemas, 1);

    ovstage_population_desc_t desc = {};
    desc.domains = OVSTAGE_POPULATION_DOMAIN_NONE;
    desc.selectors = &selector;
    desc.selector_count = 1;

    ovstage_population_enqueue_result_t pop = ovstage_population_open_usd_from_string_with_desc(
        stage_, str(kSchemaUsda), /*ordinal=*/1, /*time=*/0.0, &desc, 1);
    ASSERT_EQ(pop.status, OVSTAGE_OK);
    ASSERT_EQ(ovstage_population_wait_op(stage_, pop.op_index, OVSTAGE_TIMEOUT_INFINITE, nullptr), OVSTAGE_OK);
    // [/snippet:populate-by-schema-c]

    EXPECT_EQ(prefixCount("/World/Body"), 1u);
    EXPECT_EQ(prefixCount("/World/Prop"), 0u);
}

// Predicates compose, so one selector can narrow a selection to a branch of the stage.
TEST_F(PopulationTest, PopulateNarrowedToSubtree)
{
    // [snippet:populate-narrowed-c]
    // Several values of one kind already mean "any of these", so AND is how conditions
    // of *different* kinds combine. A nested predicate is referenced, not copied, so
    // every node has to outlive the call -- which is what the named locals are for.
    ovx_string_t types[] = { str("Cube") };
    ovx_string_t roots[] = { str("/World") };
    ovstage_population_prim_predicate_t parts[2] = {
        ovstage_population_prim_predicate_has_type(types, 1),
        ovstage_population_prim_predicate_is_under_path(roots, 1),
    };
    ovx_string_t names[] = { str("size") };

    ovstage_population_selector_t selector = {};
    selector.prim_predicate = ovstage_population_prim_predicate_and(parts, 2);
    selector.property_predicate = ovstage_population_property_predicate_has_name(names, 1);

    ovstage_population_desc_t desc = {};
    desc.selectors = &selector;
    desc.selector_count = 1;

    ovstage_population_enqueue_result_t pop = ovstage_population_open_usd_from_string_with_desc(
        stage_, str(kSchemaUsda), /*ordinal=*/1, /*time=*/0.0, &desc, 1);
    ASSERT_EQ(pop.status, OVSTAGE_OK);
    ASSERT_EQ(ovstage_population_wait_op(stage_, pop.op_index, OVSTAGE_TIMEOUT_INFINITE, nullptr), OVSTAGE_OK);
    // [/snippet:populate-narrowed-c]

    EXPECT_EQ(prefixCount("/World/Body"), 1u);
    EXPECT_EQ(prefixCount("/World/Prop"), 1u);

    EXPECT_EQ(prefixCount("/World/Ball"), 0u);
    EXPECT_EQ(prefixCount("/Outside"), 0u);
}

File: tests/python/test_population.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: LicenseRef-NvidiaProprietary
#
# NVIDIA CORPORATION, its affiliates and licensors retain all intellectual
# property and proprietary rights in and to this material, related
# documentation and any modifications thereto. Any use, reproduction,
# disclosure or distribution of this material and related documentation
# without an express license agreement from NVIDIA CORPORATION or
# its affiliates is strictly prohibited.
#
# Public ovstage population test (USD -> ovstage): populate from inline USDA and
# query back; add/remove a USD reference; reset; and confirm a missing file fails.
# These are the tested source for the loading-usd skill's snippets. The C sibling
# is tests/c/test_population.cpp. CPU-only.

import math

import pytest

from ovstage import (Filter, FilterOp, OvstageError, PathDictionary, PopulationDomain, Predicate,
                     population)

pytestmark = pytest.mark.skipif(
    not population.available(),
    reason="libovstage was built without the ovstage population bridge",
)

# A minimal USD scene: one Cube under a World Xform.
CUBE_USDA = """#usda 1.0
(
    defaultPrim = "World"
)
def Xform "World"
{
    def Cube "Cube"
    {
        double size = 1.0
    }
}
"""

# A scene with one prim carrying a physics schema and one that does not, so a
# description that names the schema is distinguishable from one that selects everything.
SCHEMA_USDA = """#usda 1.0
(
    defaultPrim = "World"
)
def Xform "World"
{
    def Cube "Body" (prepend apiSchemas = ["PhysicsRigidBodyAPI"])
    {
        double size = 4.0
    }
    def Cube "Prop"
    {
        double size = 2.0
    }
    def Sphere "Ball"
    {
        double radius = 1.0
    }
}
def Cube "Outside"
{
    double size = 9.0
}
"""

# A self-contained, referenceable layer (sets defaultPrim so the reference composes).
REF_USDA = """#usda 1.0
(
    defaultPrim = "Ref"
)
def Xform "Ref"
{
    def Cube "Cube"
    {
        double size = 1.0
    }
}
"""


def _prefix_count(stage, prefix: str) -> int:
    query = stage.query(filter=Filter([Predicate("usd-path", FilterOp.PREFIX, [prefix])]))
    query.wait()
    count = stage.fetch_query_result(query).total_prim_count
    stage.release_query(query).wait()
    return count


def _publishes(stage, prefix: str, attribute: str) -> bool:
    """Whether ovstage discovered `attribute` on the prims under `prefix`."""
    with PathDictionary(stage) as paths:
        token = paths.intern_token(attribute)
        query = stage.query(filter=Filter([Predicate("usd-path", FilterOp.PREFIX, [prefix])]))
        query.wait()
        result = stage.fetch_query_result(query)
        found = token in set(result.attributes)
        stage.release_query(query).wait()
        return found


def test_populate_from_usda_and_query(stage):
    # [snippet:populate-and-query]
    # Populate the ovstage from an inline USDA string (the RENDERING domain
    # mirrors meshes/lights/materials/cameras). open_usd_from_string blocks until
    # the populate op completes. Then confirm the prim landed by querying it back
    # by its usd-path — queries resolve against the latest committed state, so no
    # write-floor advance is needed.
    population.open_usd_from_string(
        stage, CUBE_USDA, ordinal=1, time_code=math.nan, domains=PopulationDomain.RENDERING
    )
    query = stage.query(filter=Filter([Predicate("usd-path", FilterOp.IN, ["/World/Cube"])]))
    query.wait()
    matched = stage.fetch_query_result(query).total_prim_count
    stage.release_query(query).wait()
    # [/snippet:populate-and-query]

    assert matched == 1


def test_add_remove_usd_reference(stage):
    population.open_usd_from_string(stage, CUBE_USDA, ordinal=1, domains=PopulationDomain.RENDERING)

    # [snippet:usd-reference]
    # add_usd_reference edits the USD source only; apply_usd_changes propagates it
    # into the stage (at the ordinal you pass). Each add reserves a handle for a
    # later remove; removing again propagates the tombstone.
    handle = population.add_usd_reference_from_string(stage, REF_USDA, "/World/Props")
    population.apply_usd_changes(stage, ordinal=2)
    assert _prefix_count(stage, "/World/Props") > 0  # the referenced subtree materialized

    population.remove_usd(stage, handle)
    population.apply_usd_changes(stage, ordinal=3)
    assert _prefix_count(stage, "/World/Props") == 0  # the referenced subtree was removed
    # [/snippet:usd-reference]

    # The handle is spent: removing it again is rejected.
    with pytest.raises(OvstageError):
        population.remove_usd(stage, handle)


def test_reset_usd_and_repopulate(stage):
    population.open_usd_from_string(stage, CUBE_USDA, ordinal=1, domains=PopulationDomain.RENDERING)
    assert _prefix_count(stage, "/World") > 0

    # [snippet:reset-usd]
    # reset_usd clears the USD source; apply_usd_changes propagates the cleared
    # state into the stage. The stage stays usable afterwards — repopulating from
    # USD still works.
    population.reset_usd(stage)
    population.apply_usd_changes(stage, ordinal=2)
    assert _prefix_count(stage, "/World") == 0  # cleared before repopulating
    population.open_usd_from_string(stage, CUBE_USDA, ordinal=3, domains=PopulationDomain.RENDERING)
    # [/snippet:reset-usd]

    assert _prefix_count(stage, "/World") > 0


def test_open_missing_file_fails(stage):
    # [snippet:open-missing-file]
    # A missing/unreadable file fails the populate op. The blocking open_usd
    # raises OvstageError (the async variant surfaces it from .wait()).
    with pytest.raises(OvstageError):
        population.open_usd(stage, "/nonexistent/ovstage-does-not-exist.usda", ordinal=1)
    # [/snippet:open-missing-file]


def test_populate_by_schema(stage):
    """Populate exactly the prims carrying a schema, with the properties it declares."""
    # [snippet:populate-by-schema]
    # A description says what to populate. The prim predicate picks the prims -- here
    # anything conforming to PhysicsRigidBodyAPI, whether as its type or as an applied
    # API schema -- and the property predicate picks which of their properties are
    # published. The same schema list serves both halves.
    schemas = ["PhysicsRigidBodyAPI"]
    selector = population.Selector(
        prim_predicate=population.PrimPredicate.has_schema(*schemas),
        property_predicate=population.PropertyPredicate.declared_by_schema(*schemas),
    )
    population.open_usd_from_string_with_desc(
        stage, SCHEMA_USDA, ordinal=1, time_code=math.nan,
        descs=[population.Desc(selectors=[selector])],
    )
    # [/snippet:populate-by-schema]

    # The prim carrying the schema is populated; the one that does not is not.
    assert _prefix_count(stage, "/World/Body") == 1
    assert _prefix_count(stage, "/World/Prop") == 0


def test_populate_narrowed_to_a_subtree(stage):
    """Combine predicates: the same selection, scoped to one branch of the stage."""
    # [snippet:populate-narrowed]
    # Predicates compose with and_ / or_ / not_. Several values of one kind already
    # mean "any of these", so AND is how conditions of *different* kinds combine.
    selector = population.Selector(
        prim_predicate=population.PrimPredicate.and_(
            population.PrimPredicate.has_type("Cube"),
            population.PrimPredicate.is_under_path("/World"),
        ),
        property_predicate=population.PropertyPredicate.has_name("size"),
    )
    population.open_usd_from_string_with_desc(
        stage, SCHEMA_USDA, ordinal=1, time_code=math.nan,
        descs=[population.Desc(selectors=[selector])],
    )
    # [/snippet:populate-narrowed]

    assert _prefix_count(stage, "/World/Body") == 1
    assert _prefix_count(stage, "/World/Prop") == 1
    
    assert _prefix_count(stage, "/World/Ball") == 0
    assert _prefix_count(stage, "/Outside") == 0


def test_populate_stage_metadata(stage):
    """Stage metadata is populated onto the root prim when a description asks for it."""
    # [snippet:populate-stage-metadata]
    # Stage metadata lands on the root prim, `/`, under the `usd-metadata:` prefix. A
    # path is a field name, extended with a `:`-joined key path to reach inside a
    # dictionary field.
    population.open_usd_from_string_with_desc(
        stage, SCHEMA_USDA,
        descs=[population.Desc(stage_metadata_paths=["timeCodesPerSecond"])],
        ordinal=1, time_code=math.nan,
    )
    # [/snippet:populate-stage-metadata]

    assert _prefix_count(stage, "/") >= 1


def test_descriptions_from_several_contributors(stage):
    """Descs compose, so each library can supply its own without agreeing on one."""
    # [snippet:populate-several-descs]
    # `domains` is OR-ed across the descriptions and their selectors are combined, so a
    # host can collect one description per library and pass them together.
    renderer = population.Desc(domains=PopulationDomain.RENDERING)
    mine = population.Desc(
        selectors=[
            population.Selector(
                prim_predicate=population.PrimPredicate.has_schema("PhysicsRigidBodyAPI"),
                property_predicate=population.PropertyPredicate.declared_by_schema("PhysicsRigidBodyAPI"),
            )
        ]
    )
    population.open_usd_from_string_with_desc(
        stage, SCHEMA_USDA, ordinal=1, time_code=math.nan, descs=[renderer, mine]
    )
    # [/snippet:populate-several-descs]

    assert _prefix_count(stage, "/World/Body") == 1
    assert _publishes(stage, "/World/Body", "physics:velocity")
    assert not _publishes(stage, "/World/Prop", "physics:velocity")
